src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -65,124 +68,105 @@ public RecordToBSONWritableTransformer(@Nullable String idFieldName) {
   /**
    * Transforms given {@link StructuredRecord} to {@link BSONWritable}.
    *
-   * @param original structured record to be transformed.
+   * @param record structured record to be transformed.
    * @return {@link BSONWritable} that corresponds to the given {@link StructuredRecord}.
    */
-  public BSONWritable transform(StructuredRecord original) {
-    StructuredRecord record = Strings.isNullOrEmpty(idFieldName) || DEFAULT_ID_FIELD_NAME.equals(idFieldName) ?
-      original : createRecordWithId(original);
-    return new BSONWritable(extractRecord(record, DEFAULT_ID_FIELD_NAME));
-  }
-
-  private StructuredRecord createRecordWithId(StructuredRecord original) {
-    Schema.Field idField = original.getSchema().getField(idFieldName);
+  public BSONWritable transform(StructuredRecord record) {
+    DBObject extracted = extractRecord(null, record);
+    if (Strings.isNullOrEmpty(idFieldName)) {
+      return new BSONWritable(extracted);
+    }
 
+    Schema.Field idField = record.getSchema().getField(idFieldName);
     if (idField == null) {
       throw new UnexpectedFormatException(String.format("Record does not contain identifier field '%s'.", idFieldName));
     }
-
-    Schema.Field defaultIdField = original.getSchema().getField(DEFAULT_ID_FIELD_NAME);
-    if (defaultIdField != null) {
+    if (!DEFAULT_ID_FIELD_NAME.equals(idFieldName) && record.getSchema().getField(DEFAULT_ID_FIELD_NAME) != null) {
       throw new UnexpectedFormatException(String.format("Record already contains identifier field '%s'.",
                                                         DEFAULT_ID_FIELD_NAME));
     }
 
-    List<Schema.Field> copiedFields = Objects.requireNonNull(original.getSchema().getFields()).stream()
-      .filter(field -> !idFieldName.equals(field.getName()))
-      .collect(Collectors.toList());
-
-    List<Schema.Field> recordWithIdFields = new ArrayList<>();
-    recordWithIdFields.add(Schema.Field.of(DEFAULT_ID_FIELD_NAME, idField.getSchema()));
-    recordWithIdFields.addAll(copiedFields);
-
-    String recordName = Objects.requireNonNull(original.getSchema().getRecordName(), "Record name can not be empty");
-    Schema recordWithIdSchema = Schema.recordOf(recordName, recordWithIdFields);
-    StructuredRecord.Builder builder = StructuredRecord.builder(recordWithIdSchema);
-    for (Schema.Field field : original.getSchema().getFields()) {
-      if (idFieldName.equals(field.getName())) {
-        builder.set(DEFAULT_ID_FIELD_NAME, original.get(idFieldName));
-        continue;
-      }
-
-      builder.set(field.getName(), original.get(field.getName()));
-    }
-
-    return builder.build();
+    extracted.removeField(idFieldName);
```

Comment:
instead of removing and adding, I think it's better to add a nullable 'idField' argument to extractRecord that specifies which field is the id field.

When going through the fields, you can use the field schema instead of a bunch of instanceof checks.

Reply:
Done.